docs/melange-book/counter.md

```diff
@@ -0,0 +1,176 @@
+we must wrap strings inside function calls to convert them to the
+`React.element` type. This is exactly what the `React.string` function does--if
+you hover over it, you'll see that it displays the type `string =>
+React.element`. The `make` function itself has the type `unit => React.element`,
+meaning it takes `()` as the only argument and returns an object of type
+`React.element`.
+
+What's with the `module` business? OCaml's
+[modules](https://cs3110.github.io/textbook/chapters/modules/modules.html) are
+somewhat like JavaScript modules, with one (of many) notable differences being
+that there can be multiples modules inside a single file. For now, you just need
+to know that all components in ReasonReact are modules.
+
+A little bit further down, we make use of the `App` component:
+
+```reason
+let node = ReactDOM.querySelector("#root");
+switch (node) {
+| Some(root) => ReactDOM.render(<App />, root)
+| None => Js.Console.error("Failed to start React: couldn't find the #root element")
+};
+```
+
+`React.querySelector("#root")` returns an `option(Dom.element)`, meaning that if
+it doesn't find the element, it returns `None`, and if it does find the element,
+it returns `Some(Dom.element)`, i.e. the element wrapped in the `Some`
+constructor. The `switch` expression (despite the name, don't confuse it for the
+JavaScript construct of the same name), and allows you to succinctly express:
+
+- If `node` is `Some(Dom.element)`, render the `App` component to the DOM
+- Otherwise if `node` is `None`, log an error message
+
+Don't worry if this doesn't quite make sense yet, we'll talk more about `option`
+and other [variant
+types](https://cs3110.github.io/textbook/chapters/data/variants.html)
+later.
+
+Let's create a counter component by creating a new file called `Counter.re` in
+the same directory, with the following contents:
+
+```reason
+[@react.component]
+let make = () => {
+  let (counter, setCounter) = React.useState(() => 0);
+
+  <div>
+    <button onClick={_evt => setCounter(v => v - 1)}> {React.string("-")} </button>
+    {React.string(Int.to_string(counter))}
+    <button onClick={_evt => setCounter(v => v + 1)}> {React.string("+")} </button>
+  </div>;
+};
+```
+
+This is a component with a single `useState` hook. It should look fairly
+familiar if you know about [hooks in React](https://react.dev/reference/react).
+Note that we didn't need to manually define a module for `Counter`, because all
+files in OCaml are automatically modules, with the name of the module being the
+same as the name of the file.
+
+Now let's modify `App` so that it uses our new `Counter` component:
+
+```reason
+module App = {
+  [@react.component]
+  let make = () => <Counter />;
+};
+```
+
+To display the number of the counter, we wrote
+`{React.string(Int.to_string(counter))}`, which converts an integer to a string,
+and then converts that string to `React.element`. But in OCaml, there's a
+better way:
```

Comment:
```suggestion
and then converts that string to `React.element`. In OCaml, there's a way
to apply a sequence of operations over some data so that it can be
read from left to right:
```

Reply:
Updated in https://github.com/melange-re/melange-re.github.io/pull/93/commits/2fbcdbeac74c78b6b871e42bab09cb7ebec86218